ignite/cmd/model/chain_serve.go

```diff
@@ -0,0 +1,298 @@
+	}
+
+	return m, nil
+}
+
+func (m ChainServe) View() string {
+	if m.error != nil {
+		return fmt.Sprintf("%s %s\n", icons.NotOK, colors.Error(m.error.Error()))
+	}
+
+	var view strings.Builder
+
+	switch m.state {
+	case stateChainServeStarting:
+		view.WriteString(m.renderStartView())
+	case stateChainServeRunning:
+		view.WriteString(m.renderRunView())
+	case stateChainServeRebuilding:
+		view.WriteString(m.renderRebuildView())
+	case stateChainServeQuitting:
+		view.WriteString(m.renderQuitView())
+	}
+
+	if m.state != stateChainServeQuitting {
+		// TODO: Add actions to copy mnemonics to clipboard in run view?
+		view.WriteString(m.renderActions())
+	}
+
+	return cliuimodel.FormatView(view.String())
+}
+
+func (m *ChainServe) setState(state uint) {
+	m.state = state
+}
+
+func (m ChainServe) updateCurrentModel(msg tea.Msg) (tea.Model, tea.Cmd) {
+	var cmd tea.Cmd
+
+	switch m.state {
+	case stateChainServeStarting:
+		m.startModel, cmd = m.startModel.Update(msg)
+	case stateChainServeRunning:
+		m.runModel, cmd = m.runModel.Update(msg)
+	case stateChainServeRebuilding:
+		m.rebuildModel, cmd = m.rebuildModel.Update(msg)
+	case stateChainServeQuitting:
+		m.quitModel, cmd = m.quitModel.Update(msg)
+	}
+
+	return m, cmd
+}
+
+func (m *ChainServe) processEventMsg(msg cliuimodel.EventMsg) (tea.Model, tea.Cmd) {
+	var cmd tea.Cmd
+
+	// When an error event is received it means there is an issue with
+	// the blockchain app's source code that the user must fix.
+	m.broken = msg.Group == events.GroupError
+
+	// UI responds to key press or mouse events by default but we use
+	// events and the events bus to interact with the UI during execution.
+	// State defines the model/view that is being displayed.
+	switch m.state {
+	case stateChainServeStarting:
+		// Start view displays status events until the blockchain is running or an
+		// error event is received in which case it displays the run view with an
+		// error traceback and waits until the issue is fixed.
+		// When the status finish event is not an error it means that the blockchain
+		// started successfully and the run view is displayed.
+		if msg.ProgressIndication == events.IndicationFinish {
+			m.setState(stateChainServeRunning)
+
+			// Prepare model for the run view
+			m.runModel, cmd = m.runModel.Update(msg)
+
+			return m, cmd
+		}
+	case stateChainServeRunning:
+		// Serve will be displayed when there is an error until the code is fixed.
+		// It will show an error traceback until the code is changed and the rebuild
+		// view is displayed.
+		if msg.Group == events.GroupError {
+			// Clear the current events to only display the error
+			m.runModel.ClearEvents()
+
+			// Prepare model for the run view
+			m.runModel, cmd = m.runModel.Update(msg)
+
+			return m, cmd
+		}
+
+		// When the received event is not an error make sure to change the broken property
+		if m.broken {
+			m.broken = false
+
+			// Clear error events
+			m.runModel.ClearEvents()
+		}
```

Comment:
`m.broken = msg.Group == events.GroupError`, so why are there two if statements here?

Reply:
The second if-statement can't be reached because we will already return right?

Reply:
I can add a comment to the code to clarify this case:
```
Run view shows account addresses, API URLs and the paths required to
have a context on the running blockchain app and waits for errors or
changes in the blockchain app source code.
If an error event is received during run it means that there is an error
in the app source code in which case the error message and traceback are
displayed until the code is fixed, or otherwise when an status event is
received it means that the code changed so the app must be rebuilt.
```

This is the part that I'm not happy with, the message processing section where the message logic is implemented when we use the UI pattern proposed by Bubbletea. I'm thinking I should explore it further to find a better implementation for it.

Reply:
I'm not sure I fully understand.  How else is `m.broken` modified?  By some internal logic?

Reply:
> I'm not sure I fully understand. How else is `m.broken` modified? By some internal logic?

I have to review it because the logic might not be right 👍

Reply:
@aljo242 thanks again for pointing it out. I simplified/cleaned event message processing in 31da854.